### PR TITLE
Array functions

### PR DESCRIPTION
**Pull request by eytan-c**
#4 
Done recursively. @Onoam hope that this works well with the memory model of Java. Let me know otherwise.

### CODE REVIEW

File: WAVLTree/src/WAVLTree.java [hunk 1]
```diff
@@ -99,8 +94,12 @@ public String max() {
 	 * if the tree is empty.
 	 */
 	public int[] keysToArray() {
-		int[] arr = new int[42]; // to be replaced by student code
-		return arr; // to be replaced by student code
+		int[] arr = new int[getRoot().siz];
```
Comment (Onoam) [inline WAVLTree/src/WAVLTree.java]:
```suggestion
		int[] arr = new int[getRoot().size];
```

File: WAVLTree/src/WAVLTree.java [hunk 2]
```diff
@@ -110,10 +109,66 @@ public String max() {
 	 * respective keys, or an empty array if the tree is empty.
 	 */
 	public String[] infoToArray() {
-		String[] arr = new String[42]; // to be replaced by student code
+		String[] arr = new String[getRoot().size];
```
Comment (Onoam) [inline WAVLTree/src/WAVLTree.java]:
```suggestion
		String[] arr = new String[root.size];
```

File: WAVLTree/src/WAVLTree.java [hunk 3]
```diff
@@ -110,10 +109,66 @@ public String max() {
...
+	 * The counter is to validate that we're adding values in the right places.
+	 * Stopping conditions - if we reach an empty tree, we return without
+	 * changing the array (hope this works like Python or JS)
+	 * Otherwiase, we first run on the left side of the root,
+	 * then add the root to the array and move the counter + 1,
+	 * then move to the right.
+	 * Complexity - We need to go through all the nodes in the tree (n)
+	 * and we do a set number of O(1) operations for each, thus O(n)
+	 */
+	public String[] infoToArrayRec(WAVLTree tree, String[] array, int[] counter) {
```
Comment (Onoam) [inline WAVLTree/src/WAVLTree.java]:
```suggestion
	private String[] infoToArrayRec(WAVLTree tree, String[] array, int[] counter) {
```

File: WAVLTree/src/WAVLTree.java [hunk 4]
```diff
@@ -110,10 +109,66 @@ public String max() {
...
+	 * The counter is to validate that we're adding values in the right places.
+	 * Stopping conditions - if we reach an empty tree, we return without
+	 * changing the array (hope this works like Python or JS)
+	 * Otherwiase, we first run on the left side of the root,
+	 * then add the root to the array and move the counter + 1,
+	 * then move to the right.
+	 * Complexity - We need to go through all the nodes in the tree (n)
+	 * and we do a set number of O(1) operations for each, thus O(n)
+	 */
+	public String[] keysToArrayRec(WAVLTree tree, String[] array, int[] counter) {
```
Comment (Onoam) [inline WAVLTree/src/WAVLTree.java]:
```suggestion
	private String[] keysToArrayRec(WAVLTree tree, String[] array, int[] counter) {
```

Comment (Onoam) [reply WAVLTree/src/WAVLTree.java]: All methods we only use internally should be private.

**Comment by Onoam** [comment]
You can take a look at some of the small changes I suggested, and decide on them. Other than those this looks great so you can merge this one afterwards :smile

File: WAVLTree/src/WAVLTree.java (hunk 4 above)
Comment (eytan-c) [reply WAVLTree/src/WAVLTree.java]: 👍

**Pull request merged**